hosts/pds3.py

```diff
@@ -0,0 +1,172 @@
+import os
+import pdsparser
+from pathlib import Path
+
+
+#===========================================================================
+def clean_comments(line):
+    """
+    Standard cleaner to handle commenting errors that could affect any host.
+    """
+
+    # Close-comment w/o open-comment
+    if line.strip().endswith('*/') and '/*' not in line:
+        line = '\n'
+
+    return line
+
+#===========================================================================
+def call_cleaners(cleaners, lines):
+    """
+    Call the given cleaner functions.
+
+    Inputs:
+        cleaners:       List of cleaner functions.
+        lines:          List of lines comprising the label.
+    """
+
+    for i in range(len(lines)):
+        for cleaner in cleaners:
+            lines[i] = cleaner(lines[i])
+
+    return lines
+
+#===========================================================================
+def clean_label(lines, cleaners=None, no_standard=False):
+    """
+    Correct known errors in PDS3 labels.
+
+    Inputs:
+        lines:          List of lines comprising the label.
+        cleaners:       List of functions that take a line as input and
+                        return a cleaned line.  By default, these functions
+                        are appended to a list of standard cleaner functions.
+        no_standard:    If True, the standard cleaners are not used.
+    """
+
+    # Call standard cleaners
+    standard_cleaners = [clean_comments]
+    if not no_standard:
+        lines = call_cleaners(standard_cleaners, lines)
+
+    # Call custom cleaners
+    if cleaners is not None:
+        lines = call_cleaners(cleaners, lines)
+
+    return lines
+
+#===========================================================================
+def find_label(filespec):
+    """Find the PDS3 label corresponding to the given filespec."""
+
+    # Construct candidate label filenames
+    spec = Path(filespec)
+    labelspecs = [
+        spec.with_suffix('.lbl'),
+        spec.with_suffix('.LBL') ]
+    if filespec.upper().endswith('.LBL'):
+        labelspecs.append(filespec)
+
+    # Check candidates; return first one that exists
+    for labelspec in labelspecs:
+        if os.path.isfile(labelspec):
+            return labelspec
+
+    # If no successful candidate, assume attached label
+    return filespec
+
+#===========================================================================
+def get_label(filespec, cleaners=None, no_standard=False):
+    """
+    Find the PDS3 label, clean it, load it, and parse into dictionary.
+
+    Inputs:
+        filespec:       File specification for the label.
+        cleaners:       List of functions that take a label line as input and
+                        return a cleaned line.  By default, these functions
+                        are appended to a list of standard cleaner functions.
+        no_standard:    If True, the standard cleaners are not used.
+    """
```

Comment:
If the default cleaner does something you might not want, then it shouldn't do it. So I recommend you remove that option.

Why do we need a list of cleaners? A single function, which can be defined inside the instrument module, ought to suffice.

Also, because the work of a cleaner function might have to operate across multiple lines, the function should take the full list of strings and return another list of strings. It isn't necessarily possible to correct a label if the lines are only handled one at a time.

Reply:
The purpose of mulitple cleaners was to be able to select which standard cleaners you wanted to use if you didn't want to use all of them.  I just pulled that stuff out and we can let the hosts do all the cleaning.  If we wind up with a lot of duplicated code, we can add functions to pds3 to do those things.